### PR TITLE
[DO NOT MERGE] expose NodeView to JavaScript

### PR DESCRIPTION
**Pull request by mikemorris**
Refs https://github.com/mapbox/node-mapbox-gl-native/issues/142, https://github.com/mapbox/node-mapbox-gl-native/pull/149

May end up not merging this, but wanted to reimplement it here so we can test different caching strategies.

### CODE REVIEW

**Comment by mikemorris** [comment]
Tagging @jfirebaugh for a review here.

File: platform/node/src/node_view.hpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-parameter"
+#pragma GCC diagnostic ignored "-Wshadow"
+#include <node.h>
+#include <nan.h>
+#pragma GCC diagnostic pop
+
+namespace node_mbgl {
+
+class NodeView : public node::ObjectWrap {
```
Comment (jfirebaugh) [inline platform/node/src/node_view.hpp]: I think this should be `public Nan::ObjectWrap` instead.

**Comment by mikemorris** [comment]
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  Couldn't create framebuffer: incomplete attachment
```
Seeing the above error on Travis Linux `API.RepeatedRender` test.

File: platform/node/src/node_view.hpp (hunk 1 above)
Comment (mikemorris) [inline platform/node/src/node_view.hpp]: HAH, yea bet that's it.

File: platform/node/test/js/map.test.js [hunk 2]
```diff
@@ -48,12 +48,12 @@ test('Map', function(t) {
         options.cancel = function() {};
         t.throws(function() {
             new mbgl.Map(options);
-        }, /Options object must have a numerical 'ratio' property/);
+        }, /Options object must have a 'view' or  numerical 'ratio' property/);
```
Comment (mikemorris) [inline platform/node/test/js/map.test.js]: Reminder to fix the double space after `or`

File: src/mbgl/map/map_context.cpp [hunk 3]
```diff
@@ -27,14 +27,13 @@
 #include <mbgl/util/string.hpp>
 
 #include <algorithm>
+#include <iostream>
```
Comment (mikemorris) [inline src/mbgl/map/map_context.cpp]: Drop this

**Comment by mikemorris** [comment]
The [`returns an error`](https://github.com/mapbox/mapbox-gl-native/blob/node-view/platform/node/test/js/map.test.js#L199-L218) test in `map.test.js` is now segfaulting here:
```
Process 65928 stopped
* thread #1: tid = 0x696542, 0x00000001031039c1 mapbox-gl-native.node`mbgl::Map::update(this=0x0000000100d0cc80, flags=<unavailable>) + 33 at map.cpp:97, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
    frame #0: 0x00000001031039c1 mapbox-gl-native.node`mbgl::Map::update(this=0x0000000100d0cc80, flags=<unavailable>) + 33 at map.cpp:97
   94   void Map::update(Update flags) {
   95       assert(view);
   96       if (flags & Update::Dimensions) {
-> 97           transform->resize(view->getSize());
   98       }
   99       context->invoke(&MapContext::triggerUpdate, transform->getState(), flags);
   100  }
```

**Comment by mikemorris** [comment]
^ Fixed the segfault above (`view` was still `nullptr`) and added better error handling (will now throw a JS error when attempting to call `map.render` before the `view` has been set or if the `view` dimensions don't match the requested render dimensions)

**Comment by mikemorris** [comment]
@lucaswoj @bsudekum This should be ready for some perf testing if you want to try building a `NodeView` cache/pool and swapping in matching views to the `NodeMap` at render time.

**Comment by jfirebaugh** [comment]
Currently it's looking like we don't need this to get decent performance, though it's possible we'll revisit in a future optimization pass.

**Pull request closed without merging**